### ISSUE TITLE
[Pipe] supporting None and non-Tensors in forward's input/output 

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#53952]
## 🚀 Feature

As suggested by @pritamdamania87 [here](https://github.com/pytorch/pytorch/pull/50693#issuecomment-780913337) splitting my attempted PR https://github.com/pytorch/pytorch/pull/50693 into an issue as it hasn't been resolved by my attempt.

The original description follows and there is already a longish discussion in the PR.

-------------------------------

I was trying to convert `transformers` [t5](https://github.com/huggingface/transformers/blob/357fb1c5d8b6a16f042f9b504f023d935086e8e5/src/transformers/models/t5/modeling_t5.py#L589) to use the new Pipe and I encountered a dozen of input args that some are Bools and some are optionally `None`, yet others tuples within tuples. You can see an example of the complex inputs it gets:

https://github.com/huggingface/transformers/blob/357fb1c5d8b6a16f042f9b504f023d935086e8e5/src/transformers/models/t5/modeling_t5.py#L600-L612

Currently `Pipe` only handles `input`/`output` which is either a single Tensor or a tuple of Tensors, so we can't use it as it stands now.

A user needs to be able to pass as a part of the `input` and `output` tuple:
1. `None` - in transformers these are passed to `forward` when the thing needs to be optionally generated by downstream layers, but is a Tensor at other times
2. Bools - these are flow control flags that the user's code needs to pass forward to the model. These aren't available during model's init.
3. tuples of Tensors within a tuple - there are several aggregators that append to a tuple, which is then returned as part of outputs - so again, some recursive introspection and `to()` is needed see below.

Bottom line, a whole bunch of variations of variables might be needed to be passed to and from the `forward` function in the pipe chain. As long as these structures can be traversed and switched `.to()` the right device and spliced where this is needed, any type of variable should be supported. Please, see below.

@pritamdamania87

**Comment by pritamdamania87** [comment]
Would love to know if this proposal would work:

1. Input and output of stages of the `Pipe` support a Tensor or a Tuple of arbitrary types (Tensors, bools, other tuples etc). 
2. The Tensors in the tuple are split across the batch dimension and fed to each microbatch.
3. The non-Tensor types in the Tuple are just replicated as is to each microbatch without any splitting. The assumption here is that we assume these are arguments that don't have anything to do with the batch dimension.
4. Note that you could have Tuple[Tensor, int, Tuple[Tensor]], we won't support handling Tensors within nested Tuples since it is very hard for PyTorch's autograd machinery to deal with such types.
5. Basically at the boundary of each microbatch we support the same types: Tensor or a Tuple of arbitrary types and just pass the output to the next layer as is.
6. Now the interesting part is how we aggregate results in the last layer. For Tensors, we just combine them across the batch dimension as we do today. For all other types, we combine them by just taking all the outputs from all microbatches and combining them in a List. Ex: if the output of minibatch1 is Tuple[Tensor, int] and the output of minibatch2 is Tuple[Tensor, int], the final output of Pipe would be Tuple[Tensor, List[int]].

> tuples of Tensors within a tuple - there are several aggregators that append to a tuple, which is then returned as part of outputs - so again, some recursive introspection and to() is needed see below.

The above mentioned approach doesn't really support this use case since as I mentioned above, it is hard to support such types in PyTorch's autograd machinery. One workaround would be to unwrap the nested tuples with Tensors into a flattened Tuple and have the Tensors at the top level.

@mrzzd @mrshenli @stas00 Thoughts on this approach?

**Comment by stas00** [comment]
In general this is a great proposal, @pritamdamania87! Thank you!

Would it make things simpler if we were to have:
```
def forward(inputs_to_chunk_on, inputs_to_pass_on):
    return outputs_to_recompose, outputs_to_gather
```

1. two input tuples and two output tuples, where the second tuples are optional
2. the Tensors in `inputs_to_chunk_on` are always chunked
3. the Tensors in `outputs_to_recompose` are always recomposed
4. whatever is in `inputs_to_pass_on` is passed on as is
5. whatever is in `outputs_to_gather` is combined as a list - definitely need to think more about this use case

I'm concerned that if we try to automatically derive which Tensors to slice on the batch dimension by just looking that it has the right first dimension we might not be able to pass data that looks similar but is not to be sliced? I need to support this with a use case.

The aggregates are hard. For example in the current codebase of `transformers` the aggregate collects different data for each sequential transformer block, and that's how it knows to update the right slot. I'm struggling to wrap my head around how these can be aggregated when used in `nn.Sequential` manner - originally this is done in a loop, so it's easy to recompose, but now we will have just a return value from the block down the road. We need to find a way to aggregate data in this new way of invoking the layers.

**Comment by pritamdamania87** [comment]
I'm assuming the forward signature you suggested referred to Pipe's forward method and might be something like this:

```
def forward(inputs_to_chunk_on: TensorOrTupleofTensors , *inputs_to_pass_on=None):
```

I'm assuming inputs_to_pass_on is actually something like *args. If so, its not clear how to actually pass these inputs to the first stage of the pipeline. For example, lets say the first stage of the model is:

```
def forward(t1: Tensor, flag: bool, param: int, t2: Tensor)
```

Then you would call Pipe's forward function as follows:

```
Pipe.forward(inputs_to_chunk=(t1, t2), inputs_to_pass=(flag, param))
```

But now how do you map the parameters appropriately to the first stage of the pipeline? Isn't it simpler if Pipe.forward just accepted this:

```
Pipe.forward(inputs=(t1, flag, param, t2))
```

>     return outputs_to_recompose, outputs_to_gather

I'm assuming this only applies to the last stage of the pipeline right? Since you only need to gather/recompose outputs in the last stage. If so, it seems awkward to ask users to rewrite the last stage of the pipeline in a special way. 


>I'm concerned that if we try to automatically derive which Tensors to slice on the batch dimension by just looking that it has the right first dimension we might not be able to pass data that looks similar but is not to be sliced? I need to support this with a use case.

The proposal doesn't decide which Tensors to slice based on batch dimension, it will slice all Tensors on batch dimension irrespective. Do you have a use case, where you want certain Tensors not to be sliced?

> The aggregates are hard. For example in the current codebase of transformers the aggregate collects different data for each sequential transformer block, and that's how it knows to update the right slot. I'm struggling to wrap my head around how these can be aggregated when used in nn.Sequential manner - originally this is done in a loop, so it's easy to recompose, but now we will have just a return value from the block down the road. We need to find a way to aggregate data in this new way of invoking the layers.

Doesn't the proposal of returning List[..] for non-Tensor parameters address aggregates? Basically you get the raw data for each microbatch and then can manually aggregate that data however you wish.

**Comment by stas00** [comment]
In my example I was referring to the first stage of the model, because the current spec requires it to have just a single input variable.

If you intend to chunk all Tensors, then no, there is no need for 2 inputs. I just thought that perhaps someone would not want this behavior for all tensors. 

For example take again an aggregate variable - it starts as `None` and then it gets built up in each stage, so we can't slice on it but we need to be able to get it back.

In general I feel it'll be hard for us to discuss this in the void.

Perhaps we take one of your tutorials and start extending it and adapt the API as we go? e.g. let's added support for non-tensors first and extend the tutorial code to work, then look at the next stage? I see you already started it https://github.com/pytorch/pytorch/issues/53437

We can use this as the next stage version of the tutorial - i.e more advanced and we can then try to deal with conditional branches and more complex input/outputs such as aggregates? How does that sound?

I just feel that my use case of `transformers` t5 is too complex, whereas your current tutorial is too simple, so perhaps building from the simple towards more complicated would be a good way to keep things progressing.

Please let me know if any of my suggestions resonate, and if not let's look at other ways.

Thank you, @pritamdamania87

**Comment by pritamdamania87** [comment]
> If you intend to chunk all Tensors, then no, there is no need for 2 inputs. I just thought that perhaps someone would not want this behavior for all tensors.

This does make sense, maybe Pipe.forward can have a signature like this so it makes sense which parameters go where:

```
Pipe.forward(inputs_to_chunk: Dict[str, TensorOrTupleOfTensors], inputs_to_pass: Dict[str, object])
```

As an example:

```
# If you have the forward function of your 1st stage as follows:

class Stage1(...)
   
  def forward(a: Tensor, b: int, c:Tensor, d:bool, e:Tensor)

# Lets say you want to only chunk tensors 'a' and 'e:

Pipe.forward(
   inputs_to_chunk={'a': torch.rand(10,10), 'e': torch.rand(10, 10)}, 
   inputs_to_pass: {'b': 10, 'c': torch.rand(10, 10), 'd': True}
)
```

**Comment by stas00** [comment]
Yes, I think that would be the most flexible approach. Thank you for the full example, @pritamdamania87. 

But I'd ask for a 2nd opinion since it is a somewhat unusual signature ;)

I guess the return signature can't be made this clear.

**Comment by pritamdamania87** [comment]
Thinking about this a bit more, I think this might be a much simpler API that works and is backward compatible to what we have today:

```
Pipe.forward(*inputs, no_chunk: List[str])

# If you have the forward function of your 1st stage as follows:

class Stage1(...)
   
  def forward(a: Tensor, b: int, c:Tensor, d:bool, e:Tensor)

# Lets say you want to only chunk tensors 'a' and 'e:

Pipe.forward(
  torch.rand(10, 10), 
  10, 
  torch.rand(10, 10), 
  True, 
  torch.rand(10, 10),
  no_chunk=['b', 'c', 'd'],
)
```

> Perhaps we could take one of your tutorials and start extending it and adapt the API as we go? e.g. let's add support for non-tensors first and extend the tutorial code to work, then look at the next stage? I see you already started it #53437

> We can use this as the next stage version of the tutorial - i.e more advanced and we can then try to deal with conditional branches and more complex input/outputs such as aggregates? How does that sound?

> I just feel that my use case of transformers t5 is too complex, whereas your current tutorial is too simple, so perhaps building from the simple towards more complicated would be a good way to keep things progressing.

> Please let me know if any of my suggestions resonate, and if not let's look at other ways.

Agree with this overall, I did want to nail out the API though before I go into adding support for non-Tensors and what to chunk and what not to chunk.

**Comment by stas00** [comment]
Sounds good to me. Not sure about the name though. Perhaps `input_keys_not_to_chunk`? I tried a few and keep coming up with long names since the proposed variable is quite "complex".

**Comment by mrshenli** [comment]
One alternative might be sth like below. Where `NoChunk` is a standalone type that takes any pyobject and the pyobject can be retrieved by calling `NoChunk.get()`. So that we don't need to worry about mapping the `no_chunk=['a', 'b']` strings to `forward` arguments?

```python
Pipe.forward(
  NoChunk(torch.rand(10, 10)), 
  10, 
  NoChunk(torch.rand(10, 10)), 
  True, 
  torch.rand(10, 10),
  no_chunk=['b', 'c', 'd'],
)
```

**Comment by stas00** [comment]
I like your suggestion a lot, @mrshenli , since it keeps the nochunking declaration close to each argument and not as a separate definition. The latter is somewhat error-prone to mixing things up and forgetting to update the right key in `no_chunk=[...]`

**Comment by pritamdamania87** [comment]
@stas00 The stack of PRs starting at https://github.com/pytorch/pytorch/pull/55441 have implemented this proposal. Thought I'd share this with you to see if you have any early feedback for us :) Thanks again for all the feedback so far!

**Comment by stas00** [comment]
This is so awesome to hear, @pritamdamania87! Thank you for the heads up.

What's the easiest way to see what's working, other than reading through a lot of diffs? Is there an updated tutorial or even API docs that I could read to see what the final product looks like?

I haven't been experimenting with pipeline for a while now, working on other things, but may come back to it soon.

**Comment by pritamdamania87** [comment]
@stas00 If you want to try this out right away, you can patch the stack of PRs starting here: https://github.com/pytorch/pytorch/pull/55441 and build PyTorch from source locally and try out the new API. The `Pipe` class has all the necessary docs outlining the changes.

The other option would be to use the nightly binaries once the PRs land. You can also llook at the unit tests in those PRs for examples on how to use non-Tensor inputs.

**Comment by stas00** [comment]
Thank you for the guidance, @pritamdamania87

Since the first suggestion is a project of its own, I'd rather wait until it's merged. Thank you for working on it!

**Comment by pritamdamania87** [comment]
@stas00 Took a while to merge in all the PRs, but they are all landed now. Would love to have your feedback on the updated API and if there are still any gaps that we can resolve.

**Comment by stas00** [comment]
Thank you for the heads up, @pritamdamania87 

It's wonderful to know it's finally all in.

I'm currently working on totally different things, and that line of work has been suspended for now.

So I am certainly planning to get back to work on using the great improvements you added, but I don't quite know when it'll happen. I will follow up as soon as I get a chance to try it again.
